[Dynamo] Analyze triton kernels via tracing to determine mutations

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117300

This PR adds TTIR lexing and parsing in order to analyze which of the user defined triton kernel inputs are mutated.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Differential Revision: [D53165999](https://our.internmc.facebook.com/intern/diff/D53165999)

oulgen committed:
[Dynamo] Analyze triton kernels via tracing to determine mutations

[ghstack-poisoned]

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -61,88 +65,140 @@ def reset_table(self) -> None:
+        if isinstance(o, Proxy):
+            raise Exception("Proxy operation with Scalar")
+
+    def __mul__(self, o):
```
Needing to implement all these methods is not great. I wonder if there's a way to autogenerate these

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -61,88 +65,140 @@ def reset_table(self) -> None:
+        for name, impl in inspect.getmembers(tl, inspect.isfunction):
+            default_impls[name] = impl
+
+            if name == "store":
```
This approach relies on stores being the only function that can mutate a tensor pointer. I wonder how we can make this triton pin upgrade catches problems here

jansel replied:
I think there are a handful of others like atomic_add.

oulgen replied:
Good catch. I thought this op was taking a triton tensor and not a pointer but I was wrong. I'll mark all atomic ops as mutational.
https://triton-lang.org/main/python-api/triton.language.html#atomic-ops

eellison replied:
Potentially inline assembly as well ?

oulgen replied:
@eellison I already handled that on line 151

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,89 +65,136 @@ def reset_table(self) -> None:
+if has_triton():
+    import triton
+
+    for name, _ in inspect.getmembers(triton.language.core.tensor, inspect.isfunction):
```
@ezyang After thinking more about it, I decided to opt in for the for loop approach since this is only used for triton kernel tracing, the implementations are very straightforward and do not need mypy typing. Let me know what you think.

ezyang replied:
Oh, I suppose monkeypatching is a good reason to do it programmatically.

eellison reviewed (commented):
How does this work with control flow in triton ?

oulgen commented:
@eellison Triton only allows control flow via constexpr stuff and i pass those as their real values. My test cases cover that.

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds monkey patching based tracing to user defined triton kernels to identify which inputs are being mutated. We take advantage of the fact that all triton language ops return immutable kernel values except for store operation. This means that only the input to store (or indirect index of it) can be mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented:
@jansel @eellison The only problem I have encountered with this approach is if a kernel calls another kernel, we need to monkey patch the @triton.jit on the latter callee kernel but the decorator is applied at import time, so we cannot monkey patch it. We might be able to fix it by reloading/reimporting but not super excited about that.

jansel commented:
You could also have a case like:
```
@triton.jit
def foo(a):
   b = a+1
   tl.store(b, ...)
```

Or:
```
@triton.jit
def foo(a):
   b = tl.make_block_ptr(a, ....)
   tl.store(b, ...)
```

I think you could still build a correct analysis, but rather than starting from the assumption that everything is not modified and searching for stores -- it should start by assuming everything is modified then rule out modification.

Something like
1) Find all the ops that could possibly mutate (extern calls, store, atomic_add, etc)
2) See if you can determine exactly which input they mutate
3) If you can't determine what they mutate, conservatively assume everything is mutated
4) If and only if you can attribute every mutation op to a specific input, then the remaining inputs aren't mutated

jansel reviewed (changes requested):
See comment above

oulgen commented:
@jansel 
```
@triton.jit
def foo(a):
   b = a+1
   tl.store(b, ...)
```
is actually covered by my analysis. `Proxy` implements `__add__` so we say that offsets into proxy are still the same proxy. The test cases exhibit this behavior.

Once I augment my analysis with finding all the ops that could mutate, i think that will be fully correct but how could I go about finding all the such ops?

For example, `tl.make_block_pointer` is not even available in the triton documentation: https://triton-lang.org/main/python-api/triton.language.html
The only reference to it is under tutorials: https://triton-lang.org/main/getting-started/tutorials/08-experimental-block-pointer.html

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds monkey patching based tracing to user defined triton kernels to identify which inputs are being mutated. We take advantage of the fact that all triton language ops return immutable kernel values except for store operation. This means that only the input to store (or indirect index of it) can be mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented:
@jansel so the latest version of the PR handles
* mutations by store, atomic_add, atomic_cas, atomic_max, atomic_min, atomic_xchg, make_block_ptr
* bails out on inline_asm_elementwise
* indirections by x = a_ptr + <something> and x later being passed else where via dunder methods on Proxy object

jansel commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,89 +65,142 @@ def reset_table(self) -> None:
+        "make_block_ptr",
+    }
+    try:
+        # Monkey patch all triton language functions
+        for name, impl in inspect.getmembers(tl, inspect.isfunction):
```
What will happen if the user does something like:
```python
from triton.language import store

@triton.jit
def foo(x):
   store(x, 0)
```

This is usually a pitfall of monkey patching since a copy of `store` is made in the user's scope.

jansel commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,89 +65,142 @@ def reset_table(self) -> None:
+        "atomic_max",
+        "atomic_min",
+        "atomic_xchg",
+        "make_block_ptr",
```
`make_block_ptr` isn't actually a mutation op, it is just an alternate way to do indexing calculations.

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds monkey patching based tracing to user defined triton kernels to identify which inputs are being mutated. We take advantage of the fact that all triton language ops return immutable kernel values except for store operation. This means that only the input to store (or indirect index of it) can be mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented:
Fixed the block pointer support, fixed named arguments support and added more tests

@jansel For the `from triton.language import load, store` import problem, as you mentioned, the current monkey patching cannot handle it. So for the time being, I added a test to repro this and in the code, caught the exception and pessimistically bailed out of optimization. I think this should be fine for now since I assume it is very rare for people to use triton language without doing `tl.xyz`.

I do want to fix this properly, so looking ahead, I'm curious to hear how you would recommend approaching it. Off the top of my head, the only solution I can think of is using something like `importlib.reload` to reload the imports so that our monkey patches are applied. And later reloading once again to remove our monkey patches. This, I assume, might be expensive.

ezyang commented:
"What if we don't use monkey patching to do the abstract analysis"

oulgen commented:
@ezyang -- @bdhirsh and I talked about this quite a bit, it seems like there are a few options to employ here, and discussing the pros/cons of each, it seems like monkey patching is the simplest and cleanest way to implement this. There are some downsides but IMO the benefits outweigh the disadvantages. The implementation is very short and as close to future proof as we can get.

1) AST visitor based one: This will always be insufficient as it does not track flow of data very well.
2) Liveliness analysis / Sink analysis: Will be challenging to implement and make sure is correct since Triton does not follow exact Python semantics, each time triton API changes, we might run into problems.
3) Tracing with monkey patching: Identify the sinks and identify which Proxy flows into the sinks.

ezyang commented:
What if you ran Dynamo while tracing through the Triton code. Then you could "monkeypatch" the accesses to Triton library API without actually monkeypatching them for real.

jansel commented:
`importlib.reload` won't work because importing modules (especially in user code) can have side effects.  Also, you potentially could have references to both the old and new versions of the module in other parts of the code.  Imagine code like this in global scope:
```
import othermodule
othermodule.foo = bar
othermodule.somedict.setdefault("x", y)
```
If you reimport the module then `othermodule.foo` will point to the new version and `othermodule.somedict["x"]` the old version.

I'm having some trouble convincing myself this approach is correct.  Monkeypatching can be hard to reason about...

----

Another possible way to defeat monkeypatching:
```
@triton.jit
def foo(x):
   triton.language.core.store(x, 0)
```
You only monkey patched `triton.language.store` not `triton.language.core.store`, so this won't go through your patch.

oulgen commented:
@jansel I see your point. So I guess with monkey patching we can either say we only support this optimization in this narrow form, or give up all together trying to monkeypatch.

Let me explore @ezyang 's suggestion.

> What if you ran Dynamo while tracing through the Triton code. Then you could "monkeypatch" the accesses to Triton library API without actually monkeypatching them for real.

Could elaborate more? I originally started by using `symbolic_trace`. Are you suggesting going back to using symbolic trace but this time actually implement `VariableTracker`s for tl.language methods and create fake tensors or something for input tensors?

jansel commented:
IMO this should be doable with a simpler static analysis, since Triton kernels are rather restricted.  Something dynamo-based could work, but may be overkill.

ezyang commented:
Maybe the right way to do this is send a patch to Triton. If we could register a hook for every AST element, we could just have Triton call into us to do our abstract interpretation. Or maybe we can just directly implement this analysis in Triton itself.

jansel commented:
I agree, doing this on the Triton side using the actual Triton frontend seems like the way to go.

oulgen commented:
@jansel @ezyang 

Sounds good, I'll investigate how to do this in Triton. Will also double check with ptillet to make sure he is onboard with this. It looks like @aakhundov asked about this a while ago on Triton Slack and was told to use `__restrict__ ` attribute instead of implementing analysis: https://triton-lang.slack.com/archives/C01L894JD98/p1702346798420029

Since this requires changing triton, and hence waiting for pin update, it will likely take a while before we can take advantage of it internally. To unblock MRS use case, would it be reasonable to land this PR since it is only an optimization and back it out when native triton solution is ready to be incorporated?

jansel commented:
`__restrict__` seems unrelated.... it is for saying that `ptr_a` won't overlap with `ptr_b` in memory.  I think we are just assuming won't happen here.  Functionalization should know if that is happening within our context.  I think we are looking for something much simpler from Triton.

aakhundov commented:
@oulgen @jansel @ezyang Thanks for working on this and the great discussion (which I've unfortunately missed). Couple of thoughts:

1. I agree with @jansel and @ezyang that exposing the results of some simple static analysis from under the hood of Triton and quickly checking that in PT2 seems like a more robust way than monkey-patching.

2. To clarify, the question I've posed back then in the Triton Slack was more general: "how to make sure that inputs are / aren't mutated?" And then, of course, it's quickly fallen into all kinds of edge case discussions. I agree with @jansel that, given the specificity of our use case (e.g., that PT2 does functionalization etc.), we need something (much) simpler than handling the general case.

3. I'd second @oulgen's suggestion to massage and land this PR to unblock the internal MRS use cases, until we have a more general / robust solution from Triton + using it in our code. But that only makes sense, if (in this PR) we can be conservative and assume things are mutated unless we can be 100% sure they aren't. Is the latter possible?

eellison commented:
If we're looking to unblock mrs kernel, can we use explicit user annotations in the config meta like we emit for inductor kernels: `mutated_arg_names` ?

oulgen commented:
> But that only makes sense, if (in this PR) we can be conservative and assume things are mutated unless we can be 100% sure they aren't

This is already the case. The only downsides of this PR is it bails out when things are imported ahead of time.

> can we use explicit user annotations in the config meta

This feels not great since if the user incorrectly marks something we would emit bad code. I still think this PR is an okay solution until we can build the proper solution since this PR will handle the MRS kernels as they are. To be clear, I do agree that we should build the proper solution going forward.

jansel commented:
So looking quicky at the Triton source code, the key function here is `ast_to_ttir` (
https://github.com/openai/triton/blob/597ffcf6193c6f833df7fb10f3be25979ca22cda/python/triton/compiler/code_generator.py#L1192 ).

As far as I can tell, Triton is actually doing a `ast.NodeVisitor`-based approach where it is walking the Python AST without actually executing it.  Ideally our analysis would do the same thing.  We could possible just call `ast_to_ttir` then check the `ttir`. 

Alternately, we could patch the `CodeGenerator` object in Triton to do our analysis while reusing the Triton frontend.   `tl.store` calls `tl.semantic.store` under the hood, so maybe we could leave the AST parsing part the same and intercept at the tl.semantic layer.  I'd be ok with some monkey patches to Triton to do this, provided it happens a layer where the surface area is much smaller and isn't creating a new Triton frontend.

oulgen commented:
@jansel sounds good, I'll spend this week understanding the triton internals to explore all viable options and summarize them here later.

ezyang commented:
@oulgen and I chatted about the current status. It sounds like:

* Triton team's current stance is that to do this analysis properly, we should link against Triton's MLIR library and do an out-of-tree analysis. This substantially would complicate our binary packaging story and I agree with @oulgen that the cost-benefit for doing this is not here.
* The motivation behind the mutation analysis is to catch some persnickety internal users who are completely unwilling to modify their code in any way, it's mostly just a nice to have otherwise.
* If the monkeypatching failed to work, we will correctly simply just fail the analysis and make a conservative assumption. So this isn't a silent correctness problem.
* The logic here is well encapsulated, and could be easily be replaced/improved in the future.
* I agree with @oulgen's assessment that there isn't an obvious alternative that satisfies all the requirements and is obviously simple

Given these facts, I think the right call is to just land the working code as is, and move on. The situation is not irreversible and we can reassess if the monkeypatching ends up being a bug farm. wdyt @jansel

aakhundov commented:
@ezyang Thanks for summarizing this, I second everything. Maybe just a small clarification on this part:

> The motivation behind the mutation analysis is to catch some persnickety internal users who are completely unwilling to modify their code in any way, it's mostly just a nice to have otherwise.

In my opinion, the implications of having / not having this feature a slightly wider. I.e., to use Triton kernels seamlessly in PT Eager, the users must not add any specific annotation to the kernel parameters to mark what is input vs. output. And if the USP of PT2 is "if your model works in PT Eager, just `torch.compile` it and it should work", we likely don't want to ask users to add annotations to their Triton code specifically for `torch.compile`-ing it. The pass in this PR will eliminate a lot of overhead coming from unnecessarily cloning custom Triton kernel arguments before every kernel launch, and won't require any code modifications from user side (which, IIUC, is one of the reasons why folks don't quite like TorchScript and like PT2). Does this make sense?

jansel reviewed (approved):
I still don't like this approach, but accepting to unblock things.

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds monkey patching based tracing to user defined triton kernels to identify which inputs are being mutated. We take advantage of the fact that all triton language ops return immutable kernel values except for store operation. This means that only the input to store (or indirect index of it) can be mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,89 +65,152 @@ def reset_table(self) -> None:
+
+            setattr(tl, name, fn)
+
+        kernel.fn(**proxy_kwargs)
```
> Triton only allows control flow via constexpr stuff and i pass those as their real values.

This isn't true, kernels can have dynamic control flow. The simplest example is different behavior depending on `tl.program_id`:
```python
if tl.program_id(0) == 0:
    # ...
elif tl.get_program_id(0) == 1:
   # ...
```
In which case you need to be able to visit both sides of the `if`.

I think in this case you would probably raise a type error when converting `Proxy` to a boolean so I guess this is fine though (?)


> Triton team's current stance is that to do this analysis properly, we should link against Triton's MLIR library and do an out-of-tree analysis.

tbh since this is such a simple analysis you should be able to get away with a string search on the text form of the triton IR which triton already lets you access from a compiled kernel.

oulgen replied:
Hmm, I wasn't aware of dynamic control flow. Just tried some code with it and TTIR is
```
    %15 = scf.if %14 -> (tensor<16xf32>) {
      %18 = arith.addf %9, %12 : tensor<16xf32>
      scf.yield %18 : tensor<16xf32>
    } else {
      %18 = arith.mulf %9, %12 : tensor<16xf32>
      scf.yield %18 : tensor<16xf32>
    }
```

With the current implementation, tl.program_id(0) would return Scalar(), Scalar() == 0 would also return scalar, so `if Scalar()` would be true, so we would always look at the first branch.

I can add more bail out conditions here to handle this but I wonder @peterbell10 are there more defined conditions for dynamic control flow in Triton?

aakhundov replied:
Although I haven't seen this example in the internal kernels yet, it doesn't sound that contrived. E.g., IIRC, Inductor-generated foreach kernels have this structure of multiple branches relying on `program_id` ranges.

Is it possible to harvest dependencies from both branches? Just to cover all possible paths in the "graph" from mutating ops back to inputs.

oulgen replied:
That would require analysis over the kernel but the current approach employs trace execution. I added a bail out if we cast Scalar to bool and added a test case for it.

I think the best way to do this analysis is still over the TTIR or something lower, and we should negotiate with Phil etc to get this into to Triton.

peterbell10 replied:
> I wonder @peterbell10 are there more defined conditions for dynamic control flow in Triton?

In triton, fairly arbitrary control flow is possible with `if` and `while` loops. The only restriction being that the condition must evaluate to a scalar value, but you could for example get to that scalar via a reduction from an intermediate tensor. You could also load a value from memory and branch on that.

jansel replied:
This is a good point, this analysis would be incorrect on inductor-generated foreach kernels.

oulgen replied:
Yep, yesterday, I decided against continuing with this approach after learning about dynamic control flow. I was under the impression that control flow was only via `tl.constexpr`. I'm exploring the TTIR based solution, we will need something like [MLIR Python Bindings](https://mlir.llvm.org/docs/Bindings/Python/) and a whitelist TTIR ops that write to memory, and implement the analysis ourselves.

Are we ok taking a dependency on these MLIR python bindings?

aakhundov replied:
If we somehow gain access to the Triton dialect op specs, we won't need to maintain the white list? But then I'm not sure what exactly the "TTIR based solution" would be: parsing the TTIR dump that we already have or integrating our own MLIR pass (in C++, in- or out-of-tree) into the Triton MLIR machinery.

oulgen replied:
> If we somehow gain access to the Triton dialect op specs

We won't get this easily so for the purpose of this, I'll assume we dont have it

> I'm not sure what exactly the "TTIR based solution" would be

We can `ast_to_ttir` this gives us a module that we cannot inspect. Then we call `dump()` on this and get string TTIR. We parse this with mlir python bindings and then we are free to do our own analysis.

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds monkey patching based tracing to user defined triton kernels to identify which inputs are being mutated. We take advantage of the fact that all triton language ops return immutable kernel values except for store operation. This means that only the input to store (or indirect index of it) can be mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,89 +65,155 @@ def reset_table(self) -> None:
+    # Replace tensor pointers with Proxy object. Do not replace the other
+    # inputs since tl.constexprs can be inspected by the kernel.
+    proxy_kwargs = {
+        key: (Proxy() if isinstance(value, Tensor) else value)
```
Here you assume the argument is either a `Tensor` or a constant. Could we get a dynamic value such as a `SymInt` here?

peterbell10 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,89 +65,155 @@ def reset_table(self) -> None:
+                # side effect.
+
+                def fn(*args, **kwargs):  # type: ignore[misc] # different conditional signatures
+                    return Scalar()
```
What about something like
```python
ptr = tl.where(cond, ptr1, ptr2)
tl.store(ptr ...
```

I think in general it would be safer to assume that functions propagate any pointers taken as argument, and have an exception for `tl.load`.

oulgen replied:
Correct me If I am wrong but `tl.where` does not operate over tensor pointers, but rather tensor values (https://triton-lang.org/main/python-api/generated/triton.language.where.html)

peterbell10 replied:
Essentially everything in triton is a "tensor", for example `ptr1` is a scalar tensor and `ptr1 + tl.arange(0, BLOCK)` is a rank 1 tensor of pointers.

oulgen replied:
Okay, this makes sense. I'll update the algo, thanks, nice catch!

oulgen replied:
That makes sense. Just tested on a simple case. I wonder whether tl.where is the only special operation or do we need to pessimize all analysis because a function could take two pointers and only read one propagate the other.

oulgen replied:
The part about dynamically branching based on a tensor value pretty much makes static analysis be very pessimistic even if we were to analyze TTIR the original way I suggested.

aakhundov replied:
> The part about dynamically branching based on a tensor value pretty much makes static analysis be very pessimistic even if we were to analyze TTIR the original way I suggested.

Theoretically, I agree. But I'd suggest looking through a few internal Triton kernels to get a practical sense. My hunch is that, for the majority (if not all) internal Triton kernels, the static analysis should yield good results.

jansel reviewed (changes requested):
Requesting changes since you said you plan to go with a more robust approach.

oulgen commented:
@jansel @eellison @ezyang @aakhundov @peterbell10

So bit of an update from the weekend of MLIR education:

In light of dynamic conditionals in triton, I no longer believe that execution tracing based approach is viable but rather we should do analysis based approach. I also do not think monkey patching triton internals is going to be long term sustainable so I propose analyzing the TTIR since that seem to be the cleanest and most stable approach.

Using triton's `ast_to_ttir` function, I can get the TTIR however this only has what Triton's library exposes to us, assuming we cannot change this, the only way for us to get the TTIR is by calling `dump()` or `__str__`. Later, we need to parse this TTIR into something we can analyze. There are 3 possible ways to do this:
1) Use LLVM's parsing library: e.g. https://mlir.llvm.org/docs/Bindings/Python/ This would require us to link against llvm which is likely a binary size issues, so nope.
2) Use third party library: e.g. https://github.com/spcl/pymlir I tried this one and it doesnt really work for more recent MLIR syntax.
3) Write our own lightweight Lark based MLIR parser: This one might sound scary but since we are aiming for a very simple subset of the MLIR we can write a simple grammar and have Lark parse it. I actually think this will be easy since we dont really care about the traits etc.

I wanted to hear what everyone thinks. I am personally leaning towards option 3 writing simple parser so that we do not depend on any external tool changes.

peterbell10 commented:
I wonder if you might be able to do something with monkey-patching [`triton.language.semantic`](https://github.com/openai/triton/blob/main/python/triton/language/semantic.py). The functions in this module are called by `triton.language.core` to build the actual triton IR. If you were to wrap the `ir.builder` argument with a custom class then you could walk the IR structure as the compiler builds it. This would defer AST handling to the triton compiler, and avoid the need to parse IR as well.

oulgen commented:
@peterbell10 I did look at that last week a little since @jansel also suggested it but it seemed difficult to handle control flow and possibly not very future proof. I did also run into some issues with having to handle @builtin attribute's error checking.

peterbell10 commented:
> I did look at that last week a little since @jansel also suggested it

IIUC @jansel's suggestion was to intercept `tl.semantic` calls and handle its interface, but I'm suggesting intercepting the `builder` argument to `tl.semantic` and handling the interface of `ir.builder` which is lower level and more or less a 1-1 mapping with the IR format.

I will say though, that I now realize the control flow methods call the builder directly instead of going through `tl.semantic`, so injecting the custom builder by monkey patching `tl.semantic` wouldn't be enough. You would actually would need to inject it into the `builder` attribute of the `CodeGenerator` instance that's being used to build the IR. Something like:

```python
OriginalCodeGenerator = triton.compiler.CodeGenerator
class CustomCodeGenerator(OriginalCodeGenerator):
    def __init__(*args, **kwargs):
        super().__init__(*args, **kwargs)
        self.builder = MutationAnalysisWrapper(self.builder)
```

and then monkey patch that in before calling `ast_to_ttir`.

oulgen commented:
Grepping for self.builder in https://github.com/openai/triton/blob/e6e5d5468e92ed3af3e40babdd55c3da506ab01f/python/triton/compiler/code_generator.py#L1192

there functions that do control flow like `set_insertion_point_to_end` or create new regions `create_branch`. It is possible tricker than just a 1:1 mapping but I'm sure there's a way to do it.

This being said, I am still not a big fan of monkey patching here due to maintance burden. I still think MLIR parsing will be far easier.

jansel commented:
Sounds like @peterbell10 and I are both thinking along the same direction.  I still think that will be the easiest approach, and doesn't require adding new dependencies to PyTorch.

I agree with @peterbell10 that the builder insertion point is easier.  My idea there was less about the details of `tl.semantic` and more about reusing the Triton frontend parsing and intercepting below that.

We can completely ignore `set_insertion_point_to_end` and `create_branch` in this analysis.  The order of the operations doesn't matter.  Just construct a graph of operations, and check if there is path from the input to the first arg of `tl.store()`.  Furthermore, I don't think there is an analysis we need in the future that will require these.

oulgen commented:
Okay, I will give this a try and see how it turns out. I assume I still want to call `ast_to_ttir` myself and just ignore the results.

oulgen commented:
@jansel @peterbell10 

So I have tried implementing your suggestion via
```
    class MutationAnalysisWrapper:
        def __init__(self, builder):
            self.builder = builder

        def __getattr__(self, name):
            def generic_indirection(name, *args, **kwargs):
                if name not in {"get_loc", "set_loc"}:
                    print(name, args, kwargs)
                return getattr(self.builder, name)(*args, **kwargs)

            return functools.partial(generic_indirection, name)

    OriginalCodeGenerator = triton.compiler.code_generator.CodeGenerator
    class CustomCodeGenerator(OriginalCodeGenerator):
        def __init__(self, *args, **kwargs):
            super().__init__(*args, **kwargs)
            self.builder = MutationAnalysisWrapper(self.builder)

    triton.compiler.code_generator.CodeGenerator = CustomCodeGenerator
```
I am able to intercept all the builder functions and get which `create_xyz` functions are called. However, the arguments to these are `triton._C.libtriton.triton.ir.value` looking at this via `dir()` there's no way to actually access their content without changing triton internals and exposing more functionality which i assume is more pain.

Any recommendations on how to continue this way or should i revert back to custom MLIR parsing idea?

peterbell10 commented:
> I am able to intercept all the builder functions and get which create_xyz functions are called. However, the arguments to these are `triton._C.libtriton.triton.ir.value`

Yes, but those IR values are created through the `ir.builder` interface, so at creation you can wrap those IR values with your own context data and it will be propagated throughout the IR building process. 

> Technically we can track things via their pointer value but there's no way to tell which input turns into which pointer.

`builder.get_or_insert_function` is the entry point for creating a function:
https://github.com/openai/triton/blob/a5e81f9eeb2a726214b429226c17dda00d947988/python/triton/compiler/code_generator.py#L352-L353

and that function's return value `fn` has an method `fn.args` which provides the IR values for the arguments:
https://github.com/openai/triton/blob/a5e81f9eeb2a726214b429226c17dda00d947988/python/triton/compiler/code_generator.py#L369

oulgen commented:
@peterbell10 I have got pretty much everything build including the analysis. There's one last hurdle I have not been able to get past though and I was wondering if you had any opinions. 
When `triton._C.libtriton.triton.ir.function.add_entry_block` is called, we are able to tell which parameter gets converted to `triton._C.libtriton.triton.ir.block_argument`, so I can create a mapping. However, later when these parameters are used, their object address is different. 

For example, for the `add_kernel` example on triton website, I can see that
```
 0x7f4bc73d1370 => param 0
 0x7f4bc71c9230 => param 1
 0x7f4bc71c91f0 => param 2
```
but later when create_splat is called on param 2, the block_argument input is `0x7f4bc71c95f0`. Looking at the MLIR, they all appear to be in the same block, so I'm not sure why the pointers change.

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds monkey patching based tracing to user defined triton kernels to identify which inputs are being mutated. We take advantage of the fact that all triton language ops return immutable kernel values except for store operation. This means that only the input to store (or indirect index of it) can be mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,176 @@ def reset_table(self) -> None:
+                    pass
+                else:
+                    pass
+                return result
```
There isn't a 1:1 mapping between python wrapper objects and underlying C++ objects, a new wrapper is created every time it's returned from C++. This means your dict lookup will never match. You need to create your own python wrapper object and return those from the builder, not the original pybind11 wrapper objects.

oulgen replied:
For operations over non block arguments, it seems to always be the same pointer but it does not match for block argument.

I don't think I am following the new wrapper train of thought.
If I were to create
```
class CustomIRBlockArgument(triton._C.libtriton.triton.ir.block_argument):
   ...
```
and do `if isinstance(result, triton._C.libtriton.triton.ir.block_argument)`, wrap it with `CustomIRBlockArgument`. What difference would that make?

I have also tried
```
triton._C.libtriton.triton.ir.block_argument = CustomIRBlockArgument
```
so that I can overwrite it but that did not work, i assume something around pybind issues.

aakhundov replied:
Wondering if `pybind11` wrapper object is opaque to us, or we could extract the pointer (or other identifier) of the underlying C++ object which doesn't change? Unless this sounds too low-level.

peterbell10 replied:
> I don't think I am following the new wrapper train of thought.

If I were implementing this I would use composition
```python
@dataclass
class CustomFuncOp:
     underlying: triton._C.libtriton.triton.ir.function
     ...
```

Create wrappers like these for each of the types defined in triton's `ir.cc` file. Then we have complete control over the interface without needing to monkey patch anything. All the state we care about stays entirely in python and potentially you could forego the `underlying` part entirely and just stub out everything that's not relevant to the analysis.

> I guess i am still not sure why we did not do the custom MLIR parser since that was much more trivial than this but I'm willing to see this approach through.

It seems to me that with this method skip parsing and goes directly to traversing the tree of operations which is what we actually care about. It's possible my estimation is off though and perhaps the representation returned by a parser would be much easier to work with.

peterbell10 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,176 @@ def reset_table(self) -> None:
+        result = self.original_add_entry_block()
+        for i, arg in enumerate(args):
+            if isinstance(arg, Tensor):
+                mappings[self.args(i)] = TensorParam(i)
```
Here I believe your mapping is from a python wrapper object returned by one call to `block.args(i)`; then the actual IR value used throughout the triton code is a second python wrapper object created by the `args` call in `OriginalCodeGenerator`. 

It is also suspicious that you're using `block.args` whereas triton is using `function.args`, it's possible those are not the same even in C++.

oulgen replied:
You might have misread something? self here is `mlir::triton::FuncOp`, so `self.args` is args function on the `ir.Function`, not `ir.Block`. Notice how this function is overwriting `triton._C.libtriton.triton.ir.function.add_entry_block`.

oulgen replied:
I am intercepting `add_entry_block` because until this function executes, ir.Function does not have `args()` setup yet.

peterbell10 replied:
Okay so in that case it must be that `function.args(i)` is returning a new wrapper object on the second call. 

I don't understand why you wouldn't just intercept `args` itself though?

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds monkey patching based tracing to user defined triton kernels to identify which inputs are being mutated. We take advantage of the fact that all triton language ops return immutable kernel values except for store operation. This means that only the input to store (or indirect index of it) can be mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds TTIR lexing and parsing in order to analyze which of the user defined triton kernel inputs are mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+            log.debug(
+                f"ttir analysis hit an op we do not know how to analyze: {op.name}"  # noqa: G004
+            )
+            return [key for key, value in kwargs.items() if isinstance(value, Tensor)]
```
Nit: maybe compute this once and use twice to avoid code duplication? Although this may not be required, it looks very cheap to compute.

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+        if isinstance(arg, Param):
+            mutated[arg.idx] = True
+        elif isinstance(arg, Intermediate) and not arg.fake():
+            stack = [*stack, *ops[arg].args]
```
```suggestion
            stack.extend(ops[arg].args)
```

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+    # The following is an iterative DFS algorithm
+    mutated = [False] * len(kwargs)
+    while len(stack):
+        arg = stack[-1]
+        stack.pop()
```
```suggestion
        arg = stack.pop()
```

aakhundov commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+            stack.append(op.args[idx])
+
+    # The following is an iterative DFS algorithm
+    mutated = [False] * len(kwargs)
```
Maybe also keep the `visited` set? Otherwise the DFS below can take long for densely connected CFGs (not sure if we should expect those in TTIR, but there are large Triton kernels in the wild, so maybe).

aakhundov reviewed (commented):
Thanks @oulgen, looks neat!

peterbell10 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+
+        PARAM.5: "%arg" DIGIT+
+
+        INTERMEDIATE: "%" DIGIT+
```
I think you're missing multiple return values, e.g.
```
    %4:2 = "tt.reduce"(%3#0, %3#1) <{axis = 1 : i32}> ({
```

`%4:2` means there are two values returned, which are referenced as `%4#0` and `%4#1`.

This shows up in reductions, or more complicated structured control flow situations.

oulgen replied:
I will add this as a missing feature above. I also did not implement parsing for control flow yet. Just wanted to see what everyone thinks. 

The fact that TTIR can randomly emit both of the syntaxes in the same code is pretty annoying.

```
%14 = tt.addptr %13, %4 : tensor<4x!tt.ptr<f32, 1>>, tensor<4xi32>
%15 = "tt.atomic_rmw"(%14, %12, %5) <{atomic_rmw_op = 5 ...
```

oulgen commented on requirements.txt:
```diff
@@ -18,3 +18,4 @@ fsspec
 setuptools ; python_version >= "3.12"
 packaging
 optree>=0.9.1
+lark
```
I dont know a specific version we need. So until it breaks perhaps the versionless one is fine?

aakhundov replied:
I guess, yeah, especially if we end up not exposing it in the public `requirements.txt`. Current version seems to be 1.1.9: https://pypi.org/project/lark/ .

oulgen commented on test/dynamo/test_triton_kernels.py:
```diff
@@ -893,18 +893,135 @@ def f(x, y):
+                    "BLOCK_SIZE": 4,
+                    "ACTIVATION": "add_kernel",
+                },
+                # TODO(oulgen): Multiple functions is not implemented yet
```
No they emit multiple `tt.func`s

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+    # It should be possible to move this into the grammar alltogether.
+    def convert_name(token):
+        s = token.value
+        if s[0] == '"' and s[-1] == '"':
```
that's not possible, it must be a non empty value. But i can check for len > 2

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+        if len(kernel.configs) > 0:
+            # If we are autotuning, then it doesn't matter which version gets
+            # picked for tracing purposes, so lets pick the first one
+            kwargs = {**kwargs, **kernel.configs[0].kwargs}
```
If it doesn't the autotuner will will crash since it will be missing arguments?

oulgen replied:
I added a check below

aakhundov replied:
> If it doesn't the autotuner will will crash since it will be missing arguments?

I mean, in principle, Triton kernels can have default argument values for non-Tensors? And then you can skip some of those in the config spec. Although, not sure.

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+    # Replace all FakeTensor with real tensors
+    # These replacements are needed for triton's type, key and config functions
+    args: List[Any] = []
+    for a in kwargs.values():
+        if isinstance(a, (torch.SymInt, torch.SymFloat, torch.SymBool)):
+            args.append(2)
+        elif isinstance(a, FakeTensor):
+            args.append(torch.empty(2, dtype=a.dtype))
+        else:
+            args.append(a)
```
Uh.. true. I can order them properly.

eellison commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+        if isinstance(a, (torch.SymInt, torch.SymFloat, torch.SymBool)):
+            args.append(2)
+        elif isinstance(a, FakeTensor):
+            args.append(torch.empty(2, dtype=a.dtype))
```
You should check if inputs alias each other

oulgen replied:
This is a problem irrelevant of this PR, current functionalization does not handle cloning and then reassigning the clones back. I want to plan aliased inputs to the kernel all together, there's a TODO in code about this already.

aakhundov replied:
> I want to plan aliased inputs to the kernel all together, there's a TODO in code about this already.

In internal kernels, we do have multiple inputs pointing to the same memory region (slices thereof), but all those are inputs / read-only. So, not sure what "plan" means above, but I wouldn't ban this case.

oulgen replied:
We can be more creative and say that only read only inputs can be aliases. Currently if a kernel takes a and a.view(-1) as arguments, after functionalization, they will point to different things. So this is a preexisting bug.

eellison commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+        return [key for key, value in kwargs.items() if isinstance(value, Tensor)]
+
+    # Name of mutation op to mutated parameter indices
+    MUTATION_OPS = {"tt.store": [0], "tt.atomic_cas": [0], "tt.atomic_rmw": [0]}
```
Is there any particular IR reference you're using to find the list of mutation ops?

oulgen replied:
https://github.com/openai/triton/blob/7438dce18d296bb306f84e6a71ea6ee22e59535b/include/triton/Dialect/Triton/IR/TritonOps.td#L213

file contains MemWrite trait for all write operations.
Print, Assert, Store, AtomicRMWOp, AtomicCASOp are all the write operations. For our purposes we dont care about print and assert. I'll add a link.

eellison commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,235 @@ def reset_table(self) -> None:
+
+    # The following is an iterative DFS algorithm
+    mutated = [False] * len(kwargs)
+    while len(stack):
+        arg = stack[-1]
+        stack.pop()
```
Doesnt seem to handle aliasing at all, just direct mutation on directly passed in pointers.

Maybe it's better to review this part in depth when control flow is added.

oulgen replied:
For CFG, Jason was suggesting we can just convert it to straight line and analyze both sides. But yes, I'll need to deal with outputs having same name.

eellison replied:
If you have CFG, dont you need to iterate to a fixed point of aliasing for loops ?

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds TTIR lexing and parsing in order to analyze which of the user defined triton kernel inputs are mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds TTIR lexing and parsing in order to analyze which of the user defined triton kernel inputs are mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds TTIR lexing and parsing in order to analyze which of the user defined triton kernel inputs are mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on requirements.txt:
```diff
@@ -18,3 +18,4 @@ fsspec
 setuptools ; python_version >= "3.12"
 packaging
 optree>=0.9.1
+lark
```
I'm reluctant to add a new dependency to PyTorch.  I also don't think this will work when someone `pip install`'s PyTorch (which uses setup.py).

Can this be done without an external dependency?

malfet replied:
Let me clarify my comment here:
It's fine to add any kinds of optional dependencies during CI test runs (which requirments-ci.txt/requirements.txt are about), but adding something as a mandatory dependency requires a more in-depth discussion that should include such considerations as platform availability, interoperability with other packages, transitive dependencies, impact on the import time, etc

eellison replied:
If we're worried about including Lark for the binary we could use Lark to generate a stand alone parser and check that in.

oulgen replied:
Lark is under 1MB, so hopefully that's fine but yes that would always be an alternative.

jansel commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,88 +62,255 @@ def reset_table(self) -> None:
+    # %14 = tt.addptr %13, %4 : tensor<4x!tt.ptr<f32, 1>>, tensor<4xi32>
+    # tt.store %14, %12, %5 {cache = 1 : i32, evict = 1 : i32} : tensor<4xf32>
+    # %15 = "tt.atomic_rmw"(%14, %12, %5) <{atomic_rmw_op = 5 : i32, scope = 1 : i32, sem = 4 : i32}> : (tensor<4x!tt.ptr<f32, 1>>, tensor<4xf32>, tensor<4xi1>) -> tensor<4xf32>  # noqa: B950
+    grammar = """
```
How do we convince ourselves this grammar is complete?  How we we detect changes in the future?

oulgen replied:
This grammar is in fact not complete as I mentioned on the TODO at the top of the file. I'll add more rules as we need. I'm planning to work on control flow next. If the grammar changes, the tests I have added will start failing. If we fail to parse the file, we just skip on this optimization so it is always safe.

aakhundov replied:
I think, if we systematically fallback to the conservative mutation tracking if the analysis / grammar fails, we should be good?

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds TTIR lexing and parsing in order to analyze which of the user defined triton kernel inputs are mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds TTIR lexing and parsing in order to analyze which of the user defined triton kernel inputs are mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented:
Updated the PR past the triton 3.0 pin update

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds TTIR lexing and parsing in order to analyze which of the user defined triton kernel inputs are mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented:
@jansel @peterbell10

A few updates, assuming we agree on the approach, the current code is at a land-able state.

There are 3 main components:
1) `generate_ttir`: Uses triton's CodeGenerator to generate TTIR
2) `parse_ttir`: Parses this TTIR using Lark.
3) `analyze_kernel_mutations`: Determines input pointer mutations (perhaps I should rename this)

Lark is optional. If Lark is not installed, we will skip this optimization. We can get Lark can generate a stand-alone LALR parser from the grammar and check this in to avoid Lark all together but IMO this is last resort.

TTIR parser is not yet complete grammar. I will improve this with follow up PRs. Currently, if we fail to parse, we bail out on the optimization. We never incorrectly optimize anything. I left a list of missing grammar points as TODO in code. I suspect completing all of these should be quite simple.

jansel commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -60,91 +62,271 @@ def reset_table(self) -> None:
+        import lark
+        from lark import Lark, Transformer, v_args
+    except ModuleNotFoundError:
+        log.debug("Cannot parse TTIR as Lark is not installed")
```
```suggestion
        warnings.warn("Using slow path for user-defined Triton kernels. `pip install lark` to fix this.")
```

jansel reviewed (changes requested):
Do we have a test that covers the lark-not-installed case?

oulgen committed:
Update on "[Dynamo] Analyze triton kernels via tracing to determine mutations"


This PR adds TTIR lexing and parsing in order to analyze which of the user defined triton kernel inputs are mutated.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

oulgen commented:
@jansel Added a config option to disable the optimizations and duplicated the triton tests with `make_test_cls_with_patches`.

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53165999).

oulgen commented:
@pytorchbot merge
